enhancements/installer/cloud-custom-dns.md

```diff
@@ -0,0 +1,658 @@
+
+```go
+type AWSPlatformStatus struct {
+        <snip>
+        // AWSClusterDNSConfig contains all the DNS config required to configure a custom DNS solution.
+        // +optional
+        AWSClusterDNSConfig *ClusterDNSConfig `json:"awsClusterDNSConfig,omitempty"`
+
+        <snip>
+
+}
+
+type ClusterDNSConfig struct {
+        // APIServerDNSConfig contains information to configure DNS for API Server.
+        // This field will be set only when the userConfiguredDNS feature is enabled.
+        APIServerDNSConfig []DNSConfig `json:"apiServerDNSConfig,omitempty"`
+
+        // InternalAPIServerDNSConfig contains information to configure DNS for the Internal API Server.
+        // This field will be set only when the userConfiguredDNS feature is enabled.
+        InternalAPIServerDNSConfig []DNSConfig `json:"internalAPIServerDNSConfig,omitempty"`
+
+        // IngressDNSConfig contains information to configure DNS for cluster services.
+        // This field will be set only when the userConfiguredDNS feature is enabled.
+        IngressDNSConfig []DNSConfig `json:"ingressDNSConfig,omitempty"`
+}
+
+
+type DNSConfig struct {
+       // recordType is the DNS record type.
+       RecordType  string `json:"recordType"`
+
+       // lBIPAddress is the Load Balancer IP address for DNS config
+       LBIPAddress string `json:"lbIPAddress"`
+}
+
+```
+
+2. Install config is updated to allow the customer to specify if an external
+user configured DNS will be used. `UserConfiguredDNS` is added to the
+install-config and will have to be explicitly set to `Enabled` to enable this
+functionality. This config is not added to any platform specific section of
+the config because there are plans to allow this functionality in Azure and GCP
+too. The validation for this config will disallow this value being `Enabled` in
+platforms that currently do not support it.
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  annotations:
+    controller-gen.kubebuilder.io/version: v0.9.2
+  creationTimestamp: null
+  name: installconfigs.install.openshift.io
+spec:
+  group: install.openshift.io
+  names:
+    kind: InstallConfig
+    listKind: InstallConfigList
+    plural: installconfigs
+    singular: installconfig
+  scope: Namespaced
+  versions:
+  - name: v1
+    schema:
+<snip>
+      userConfiguredDNS:
+        description: UserConfiguredDNS is set to `Enabled` when the customer
+          wants to use a DNS solution external to the cluster and OpenShift is
+          not allowed to configure it. The default value is `Disabled`. Current
+          set of supported platforms is limited to AWS.
+        enum:
+        - ""
+        - Enabled
+        - Disabled
+        type: string
+```
+
+3. Add a field within the `PlatformSpec` for AWS, Azure and GCP to indicate if
+custom DNS is enabled. `PlatformSpec` is within the `Spec` field of the
+Infrastructure CR. Here is the update for platform AWS.
```

Comment:
Good point. I added it to the `Spec` field because this is configuration that is provided by the user. But, as you point out this cannot be changed after the cluster has been bootstrapped. If the `Status` field is a better place for it, please let me know.

Reply:
In general, if it cannot be changed on day 2, it should be in status. If it _can_ be changed day 2, then we tend to have a spec field that is reflected into the status once the controllers that observe the configuration have had a chance to observe and update themselves based on the new input.

I think for this case, status only is sufficient